File: java/first_opt/DBConnection.java
package myz.myDBLoading;

import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.sql.*;
import java.util.Properties;

public class DBConnection {
    private Connection connection = null;

    public DBConnection(String propPlace) {
        openDB(loadProperties(propPlace),true);
    }
    public DBConnection(String propPlace, boolean noAutoCommit) {
        openDB(loadProperties(propPlace),noAutoCommit);
    }

    private void openDB(Properties prop,boolean noAutoCommit){
        try {
            Class.forName("org.postgresql.Driver");
        } catch (ClassNotFoundException e) {
            System.err.println("Cannot find the Postgres driver. Check CLASSPATH.");
            System.exit(1);
        }

        String url = "jdbc:postgresql://" + prop.getProperty("host") + "/" + prop.getProperty("database");

        try {
            connection = DriverManager.getConnection(url, prop);
            if (connection != null) {
                System.out.println("Successfully connected to the database "
                        + prop.getProperty("database") + " as " + prop.getProperty("user"));
                if (noAutoCommit) {
                    connection.setAutoCommit(false);
                    //禁用自动提交事务 直到使用connection.commit()再提交事务
                }
            }
        } catch (SQLException e) {
            System.err.println("Database connection failed");
            System.err.println(e.getMessage());
            System.exit(1);
        }

    }

    private static Properties loadProperties(String propPlace){
        Properties prop = new Properties();
        try {
            prop.load(new InputStreamReader(Files.newInputStream(Paths.get(propPlace))));
        }catch (IOException e){
            System.err.println("can not find db user file");
            throw new RuntimeException(e);
        }
        return prop;
    }



    public void closeDB() {
        if (connection != null) {
            try {
                connection.commit();
                connection.close();
                connection = null;
                System.out.println("DataBase closed");
            } catch (Exception e) {
            }
        }
    }


    public Connection getConnection() {
        return connection;
    }

}

File: py/DbConnection.py
import configparser
import os
import psycopg2


class DBConnection:

    def __init__(self):
        self.connection = None
        self.cursor = None

    def open_db(self, config):
        try:
            parameters = {
                'host': config['DEFAULT']['host'],
                'database': config['DEFAULT']['database'],
                'user': config['DEFAULT']['user'],
                'password': config['DEFAULT']['password'],
                'port': config['DEFAULT']['port']
            }

            self.connection = psycopg2.connect(**parameters)
            self.cursor = self.connection.cursor()

            if not self.connection is None:
                print(f"Successfully connected to the database {parameters['database']} as {parameters['user']}")
                self.connection.autocommit = False

        except psycopg2.Error as e:
            raise RuntimeError("Connection Failed. ")

    def load_properties(self):
        config = configparser.ConfigParser()
        try:
            if not os.path.exists("../resources/dbUser.properties"):
                raise FileNotFoundError(f"Cannot find properties file: resources/dbUser.properties")

            config.read("../resources/dbUser.properties")
            if not config['DEFAULT']:
                raise ValueError("Properties file is empty or invalid. ")

            return config

        except (FileNotFoundError, ValueError) as e:
            print(f"Error: {e}")
            raise

    def close_db(self):
        if not self.connection is None:
            if not self.cursor is None:
                self.cursor.close()
            self.connection.close()
            self.connection = None

    def get_connection(self):
        if self.connection is None:
            raise RuntimeError("Connection not opened.")
        return self.connection

    def get_stmt(self):
        return self.cursor
